typescript/agentkit/src/action-providers/sushi/sushiRouterActionProvider.ts

```diff
@@ -0,0 +1,457 @@
+- The human-readable amount of the 'from token' to swap
+- The from token address to trade
+- The token address to receive from the swap
+- The maximum slippage allowed for the swap, where 0 is 0% and 1 is 100%, the default is 0.005 (0.05%)
+
+Important notes:
+- The native asset (ie 'eth' on 'ethereum-mainnet') is represented by ${nativeAddress} (not the wrapped native asset!)
+- Fetch a quote first before the swap action. Stop, ask the user if they want to proceed. If the user answers affirmatively, then swap
+- If you are not absolutely sure about token addresses, either use an action to fetch the token address or ask the user
+`,
+    schema: SwapSchema,
+  })
+  async swap(
+    walletProvider: ViemWalletProvider,
+    args: z.infer<typeof SwapSchema>,
+  ): Promise<string> {
+    try {
+      // Compatible chainId is expected since it should be pre-checked in supportsNetwork
+      const chainId = Number((await walletProvider.getNetwork()).chainId);
+      const chain = EvmChain.from(chainId)!;
+
+      const decimalsIn = await fetchDecimals({ walletProvider, token: args.fromAssetAddress });
+
+      if (!decimalsIn.success) {
+        return decimalsIn.message;
+      }
+
+      const amountIn = parseUnits(args.amount, decimalsIn.decimals);
+
+      // First fetch to see if the swap is even possible
+      const firstSwap = await handleGetSwap({
+        amount: amountIn,
+        chainId,
+        tokenIn: args.fromAssetAddress,
+        tokenOut: args.toAssetAddress,
+        maxSlippage: args.maxSlippage,
+        recipient: walletProvider.getAddress() as Address,
+      });
+
+      if (firstSwap.swap.status !== RouteStatus.Success) {
+        return firstSwap.message;
+      }
+
+      // Check if the wallet has enough balance to perform the swap
+      const balance = await handleBalance({
+        walletProvider,
+        token: firstSwap.swap.tokenFrom,
+        minAmount: amountIn,
+      });
+
+      if (!balance.success) {
+        return balance.message;
+      }
+
+      const approval = await handleApproval({
+        walletProvider,
+        token: args.fromAssetAddress,
+        to: firstSwap.swap.tx.to,
+        amount: amountIn,
+      });
+
+      if (!approval.success) {
+        return approval.message;
+      }
+
+      // Refetch in case the route changed during approval
+      const secondSwap = await handleGetSwap({
+        amount: amountIn,
+        chainId,
+        tokenIn: args.fromAssetAddress,
+        tokenOut: args.toAssetAddress,
+        maxSlippage: args.maxSlippage,
+        recipient: walletProvider.getAddress() as Address,
+      });
+
+      if (secondSwap.swap.status !== RouteStatus.Success) {
+        return secondSwap.message;
+      }
+
+      const swapHash = await walletProvider.sendTransaction(secondSwap.swap.tx);
+      const swapReceipt: TransactionReceipt =
+        await walletProvider.waitForTransactionReceipt(swapHash);
+
+      if (swapReceipt.status === "reverted") {
```

Comment:
User Operation from cdp smart wallets return "failed"
```suggestion
      if (swapReceipt.status === "reverted" || swapReceipt.status === "failed") {
```